### ISSUE TITLE
Duplicated cell names after concatenation

### ISSUE THREAD

**Issue by dawe** [scverse/scanpy#55]
In DropSeq experiments cell names are encoded by 12nt barcodes. It seems that no name check is performed when merging multiple datasets in ScanPy. 

```python
>>> from collections import Counter
>>> import scanpy.api  as sc

>>> f = sc.read("data1.txt").transpose()
>>> g = sc.read("data2.txt").transpose()
>>> c = f.concatenate(g)

>>> len(c.obs_names)
7932
>>> len(set(c.obs_names))
7890

>>> cc = Counter(c.obs_names)
>>> cc.most_common(10)
[('AAAAAAAAAAAA', 2), ('TCCTGTCTCTTA', 2), ('CGCAAGGGAAAG', 2), ('ACCCGTCTATGT', 2), ('CTCCTGTCTCTT', 2), ('TTCCTGTCTCTT', 2), ('CCCTGTCTCTTA', 2), ('CCGCTGTCTCTT', 2), ('GACAAACCTACC', 2), ('ACACTGTCTCTT', 2)]

```

**Comment by falexwolf** [comment]
Hi Davide,

thank you! Currently, we use the default behavior of pandas concatenate: see [here](https://github.com/theislab/anndata/blob/562954b43a9b8faa969e0ec01707bc56cbc021b0/anndata/base.py#L1371-L1400). I'll not be able to fix this during the next days. @flying-sheep, could you have a look and maybe figure out a meaningful option or meaningful default to circumvent this? It should be easy to simply pass an option to DataFrame.concat().

@dawe Thanks again for your pull request. I also put a new anndata version that incorporates it on PyPI.

Cheers, 
  Alex

**Comment by flying-sheep** [comment]
What would be a useful default?

I would assume: Drop identical observations and throw an error if observations with the same ID but different data exist.

**Comment by dawe** [comment]
I wouldn't make assumptions on the data themselves, it may be that different cells (with the same name) in different experiments get the same counts for the same gene, especially if this is low (0-2).
I'm also a Seurat user, which allows only pairwise merging of different dataset and I believe that a useful default would be add a N_ at the beginning of observation names with N being in a range(len(experiments)). This would allow seamless automatic merging of multiple datasets.

**Comment by flying-sheep** [comment]
Hmm, maybe… observation names aren’t something I commonly use to refer to observations, mostly metadata. So automatically changing them might have less impact than fiddling with any other kind of IDs.

@falexwolf, what do you think?

**Comment by dawe** [comment]
I’m working with an implementation of dropseq that needs to preserve cell names as I need to match two sequencing technologies that identify the same barcode, I’d rather preserve names (or store them in an optional dictionary)

**Comment by falexwolf** [comment]
@dawe What if you use the 'batch' field in `adata.obs` together with the index that might contain duplicates  to get the uniqueness that your're missing.

If that's not enough. I would do the same thing as [pandas.concat](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html), namely generating a multi-index.

This would be just a one-line edit for `AnnData.concatenate()` and we should do this; we simply need to make sure that the indexing behavior of AnnData as a whole remains consistent with the following. @flying-sheep, could you do this? 
```
import pandas as pd
s1 = pd.DataFrame({'v1': ['a', 'b'], 'v2': [2, 3]})
s2 = pd.DataFrame({'v1': ['c', 'd'], 'v2': [3, 4]})
s = pd.concat([s1, s2], keys=['s1', 's2'])
print('... the concatenated annotations')
print(s)
print('... all observations corresponding to "s1"')
print(s.loc['s1'])
print('... a single observation')
print(s.loc['s1', 0])
print('... values of a single observation')
# this is what we do not want in AnnData, only the behavior of the next line
# that is, a multi-index should be indexed with lists or tuples
print(s.loc['s1', 0].values)
print('... single observation and a single variable')
print(s.loc[['s1', 0], 'v1'])
print('... single observation and a single variable')
print(s.loc[('s1', 0), 'v1'])
```
gives
```
... the concatenated annotations
     v1  v2
s1 0  a   2
   1  b   3
s2 0  c   3
   1  d   4
... all observations corresponding to "s1"
  v1  v2
0  a   2
1  b   3
... a single observation
v1    a
v2    2
Name: (s1, 0), dtype: object
... values of a single observation
['a' 2]
... single observation and a single variable
s1  0    a
    1    b
Name: v1, dtype: object
... single observation and a single variable
a
```

**Comment by falexwolf** [comment]
Instead of passing a `keys` parameter to `AnnData.concatenate()`, I would pass a "multi_index" parameter, that triggers passing `keys=[0, 1, ..., n_datasets]` to the `keys` parameter of `pandas.concat`. This is then consistent with the content of the `batch` field.

**Comment by falexwolf** [comment]
So, this is solved in anndata 0.5 and scanpy 0.4.3. See the release notes (https://scanpy.readthedocs.io) and https://github.com/theislab/anndata/commit/63500075e926f202e856bd04ec673df55bbd2460 and the [example](http://anndata.readthedocs.io/en/latest/anndata.AnnData.concatenate.html).

Hope this is a meaningful default. If you pass `index_unique=None`, then it keeps the previous indices including duplicates.
